docs/cookbook/payments/custom-field-on-refund-payment.rst

```diff
@@ -0,0 +1,607 @@
+    {
+        /** @var UnitRefundTotalCalculatorInterface */
+        private $unitRefundTotalCalculator;
+
+        public function __construct(UnitRefundTotalCalculatorInterface $unitRefundTotalCalculator)
+        {
+            $this->unitRefundTotalCalculator = $unitRefundTotalCalculator;
+        }
+
+        public function fromRequest(Request $request): BaseRefundUnits
+        {
+            Assert::true($request->attributes->has('orderNumber'), 'Refunded order number not provided');
+
+            $units = $this->filterEmptyRefundUnits(
+                $request->request->has('sylius_refund_units') ? $request->request->all()['sylius_refund_units'] : []
+            );
+            $shipments = $this->filterEmptyRefundUnits(
+                $request->request->has('sylius_refund_shipments') ? $request->request->all()['sylius_refund_shipments'] : []
+            );
+
+            if (count($units) === 0 && count($shipments) === 0) {
+                throw InvalidRefundAmount::withValidationConstraint('sylius_refund.at_least_one_unit_should_be_selected_to_refund');
+            }
+
+            /** @var string $comment */
+            $comment = $request->request->get('sylius_refund_comment', '');
+
+            // here we need to return the new RefundUnits command, with new data
+            return new RefundUnits(
+                $request->attributes->get('orderNumber'),
+                $this->parseIdsToUnitRefunds($units, RefundType::orderItemUnit(), OrderItemUnitRefund::class),
+                $this->parseIdsToUnitRefunds($shipments, RefundType::shipment(), ShipmentRefund::class),
+                (int) $request->request->get('sylius_refund_payment_method'),
+                $comment,
+                new \DateTime($request->request->get('sylius_scheduled_at'))
+            );
+        }
+
+        /**
+         * Parse shipment id's to ShipmentRefund with id and remaining total or amount passed in request
+         *
+         * @return array|UnitRefundInterface[]
+         */
+        private function parseIdsToUnitRefunds(array $units, RefundType $refundType, string $unitRefundClass): array
+        {
+            $refundUnits = [];
+            foreach ($units as $id => $unit) {
+                $total = $this
+                    ->unitRefundTotalCalculator
+                    ->calculateForUnitWithIdAndType($id, $refundType, $this->getAmount($unit))
+                ;
+
+                $refundUnits[] = new $unitRefundClass((int) $id, $total);
+            }
+
+            return $refundUnits;
+        }
+
+        private function filterEmptyRefundUnits(array $units): array
+        {
+            return array_filter($units, function (array $refundUnit): bool {
+                return
+                    (isset($refundUnit['amount']) && $refundUnit['amount'] !== '') ||
+                    isset($refundUnit['full'])
+                ;
+            });
+        }
```

Comment:
Perhaps both functions could be extracted to separate service/services? Despite `InvalidRefundAmount::withValidationConstraint` they can be combined easily and mentioned exception can be placed in this service as well. Or perhaps one implementation with two services oriented on order item and shipment separated? An additional side effect of such implementation would be to hardcode classes and types in a service container instead of this class.

Reply:
I've improved this cookbook in a separate PR: https://github.com/Sylius/Sylius/pull/12748 as it waits for changes in RefundPlugin